### PR TITLE
flood control

### PR DESCRIPTION
**Pull request by avyayv**
see [this issue](https://github.com/apertium/phenny/issues/434)

### CODE REVIEW

**Comment by sushain97** [comment]
Ideally, these changes would surface more options in the config instead. For example, the config in https://github.com/apertium/phenny/pull/422#issuecomment-408887463 could be modified to support something like `'issue'` for all issue related hooks or `'issue_opened', 'issue_closed'` for only those. @Androbin made a more complete enumeration in https://github.com/apertium/phenny/issues/434#issuecomment-430340339.

File: modules/git.py [hunk 1]
```diff
@@ -260,11 +266,18 @@ def do_POST_unsafe(self, data):
...
 
-                if action == 'created':
-                    template = '{:}: {:} * comment created on {:} #{:}: {:}'
-                    messages.append(template.format(repo, user, text, number, url))
+                if event == 'issue_comment':
+                    template = '{:}: {:} * comment {:} on {:} #{:}: {:}'
+                    messages.append(template.format(repo, user, action, text, number, url))
+
+                else:
+                    what_we_want_from_event = event[14:]
```
Comment (sushain97) [inline modules/git.py]: Both this variable name and the hardcoded number are :\ Can you put some comments here to help understand what's going on?

File: modules/git.py [hunk 2]
```diff
@@ -276,9 +289,17 @@ def do_POST_unsafe(self, data):
...
-                if action == "opened" or action == "closed" or action == "reopened":
+                if event == 'issues':
                     messages.append(template.format(repo, user, number, title, action, opt, url))
 
+                else:
+                    what_we_want_from_event = event[7:]
+                    if action == what_we_want_from_event:
+                        template = '{:}: {:} * comment {:} on {:} #{:}: {:}'
+                        messages.append(template.format(repo, user, action, text, number, url))
+
```
Comment (sushain97) [inline modules/git.py]: Lots of extra newlines here.

File: modules/git.py [hunk 3]
```diff
@@ -125,6 +125,7 @@ def do_POST(self):
                 data = json.loads(post_data['payload'][0])
             else:
                 data = json.loads(post_data)
+
```
Comment (sushain97) [inline modules/git.py]: Extra line?

File: modules/git.py [hunk 4]
```diff
@@ -204,20 +205,23 @@ def do_POST_unsafe(self, data):
...
-
                 for key, value in config.git_events.items():
                     if fnmatch(full_name, key):
                         event_types = value
             else:
                 event_types = None
 
-            if (event_types is not None) and (event not in event_types):
+            event_specification_exists = False
+            for variable_type in event_types:
```
Comment (sushain97) [inline modules/git.py]:
```suggestion
            for event_type in event_types:
```

File: modules/git.py [hunk 5]
```diff
@@ -204,20 +205,23 @@ def do_POST_unsafe(self, data):
...
+                if event in variable_type:
+                    event = variable_type
+                    event_specification_exists = True
+
+            if (event_types is not None) and ((event not in event_types) and (not event_specification_exists)):
                 return [], []
 
             if config.git_channels:
                 full_name = data['repository']['full_name']
-                channels = []
```
Comment (sushain97) [inline modules/git.py]: Why was this deleted?

File: modules/git.py [hunk 6]
```diff
@@ -204,20 +205,23 @@ def do_POST_unsafe(self, data):
...
-
```
Comment (sushain97) [inline modules/git.py]: Why was this spacing deleted? Same question below.

File: modules/git.py [hunk 7]
```diff
@@ -237,50 +240,63 @@ def do_POST_unsafe(self, data):
                         data['comment']['body'],
                         template.format(repo, user, action, commit, '{}', url)
                     ))
+
```
Comment (sushain97) [inline modules/git.py]: Spacing?

File: modules/git.py [hunk 8]
```diff
@@ -237,50 +240,63 @@ def do_POST_unsafe(self, data):
...
+                    """
+                    Since format is in the following: 'issue_comment_x' and we
+                    want to isolate the x, we just remove the first 14 characters,
+                    or issue_comment_, thus the 'event[14:]'
+                    """
+                    specification = event[14:]
+                    if action == specification:
+                        template = '{:}: {:} * comment {:} on {:} #{:}: {:}'
+                        messages.append(template.format(repo, user, action, text, number, url))
+
```
Comment (sushain97) [inline modules/git.py]:
```suggestion
```

File: modules/git.py [hunk 9]
```diff
@@ -237,50 +240,63 @@ def do_POST_unsafe(self, data):
...
-                        data['comment']['body'],
-                        template.format(repo, user, action, text, number, '{}', url)
-                    ))
-            elif event == 'issues':
+                    """
+                    Since format is in the following: 'issue_comment_x' and we
+                    want to isolate the x, we just remove the first 14 characters,
+                    or issue_comment_, thus the 'event[14:]'
+                    """
+                    specification = event[14:]
```
Comment (sushain97) [inline modules/git.py]: Why not just split on `_` and take `-1`?

File: modules/git.py [hunk 10]
```diff
@@ -237,50 +240,63 @@ def do_POST_unsafe(self, data):
...
                     opt += 'with ' + data['label']['name']
+                if event == 'issues':
+                    messages.append(template.format(repo, user, number, title, action, opt, url))
+                else:
+                    """
+                    Since format is in the following: 'issues_x' and we
+                    want to isolate the x, we just remove the first 7 characters,
+                    or issues_, so we just get x, thus the 'event[7:]'
+                    """
+                    specification = event[7:]
```
Comment (sushain97) [inline modules/git.py]: Same split thing here.

Comment (sushain97) [inline modules/git.py]:
```suggestion
                    config_action = event[7:]
```

File: modules/git.py (hunk 9 above)
Comment (sushain97) [inline modules/git.py]:
```suggestion
                    config_action = event[14:]
```

File: modules/git.py [hunk 11]
```diff
@@ -204,20 +205,23 @@ def do_POST_unsafe(self, data):
...
                 event_types = []
-
                 for key, value in config.git_events.items():
                     if fnmatch(full_name, key):
                         event_types = value
             else:
                 event_types = None
 
-            if (event_types is not None) and (event not in event_types):
+            event_specification_exists = False
```
Comment (sushain97) [inline modules/git.py]:
```suggestion
            event_in_config = False
```

or something like that. Let's not use both the word spec and config.

File: modules/git.py [hunk 12]
```diff
@@ -204,20 +204,24 @@ def do_POST_unsafe(self, data):
             if config.git_events:
                 full_name = data['repository']['full_name']
                 event_types = []
-
```
Comment (sushain97) [inline modules/git.py]: Spacing?

File: modules/git.py (hunk 6 above)
Comment (sushain97) [reply modules/git.py]: Don't see why this was resolved.

File: modules/git.py [hunk 13]
```diff
@@ -227,7 +231,6 @@ def do_POST_unsafe(self, data):
                 url = data['comment']['html_url']
                 url = url[:url.rfind('/') + 7]
                 action = data['action']
-
```
Comment (sushain97) [inline modules/git.py]: Spacing

File: modules/git.py [hunk 14]
```diff
@@ -242,45 +245,54 @@ def do_POST_unsafe(self, data):
                 ref = data['ref']
                 type_ = data['ref_type']
                 messages.append(template.format(repo, user, type_, ref, event))
+
```
Comment (sushain97) [inline modules/git.py]: Spacing

File: modules/git.py [hunk 15]
```diff
@@ -242,45 +245,54 @@ def do_POST_unsafe(self, data):
...
             elif event == 'fork':
                 template = '{:}: {:} forked this repo {:}'
                 url = data['forkee']['html_url']
                 messages.append(template.format(repo, user, url))
-            elif event == 'issue_comment':
+
```
Comment (sushain97) [inline modules/git.py]: Spacing

File: modules/git.py [hunk 16]
```diff
@@ -242,45 +245,54 @@ def do_POST_unsafe(self, data):
...
+                    want to isolate the x, we split at the _ and get the last part'
+                    """
+                    config_action = event.split('_')[-1]
+                    if action == config_action:
+                        template = '{:}: {:} * comment {:} on {:} #{:}: {:}'
+                        messages.append(template.format(repo, user, action, text, number, url))
+
+            elif 'issues' in event:
                 template = '{:}: {:} * issue #{:} "{:}" {:} {:} {:}'
-
```
Comment (sushain97) [inline modules/git.py]: Spacing

**Comment by avyayv** [comment]
Example:
`
git_events = {
    '*': ['push'],
    'apertium/*': ['issues', 'push'],
    'apertium/apertium-all': [],
    'apertium/apertium-incubator': [],
    'apertium/apertium-nursery': [],
    'apertium/apertium-staging': [],
    'apertium/apertium-trunk': [],
    'apertium/apertium-tools': [],
    'avyayv/greatcheaprepowithnocode': ['issues_opened', 'issues_closed', 'issues_reopened', 'push', 'issue_comment_created', 'issue']
}
`
When we add a comment, begiak says:
`greatcheaprepowithnocode: avyayv * comment created on issue #8: https://github.com/avyayv/greatcheaprepowithnocode/issues/8`

However, when we edit a comment, nothing happens, because we specified so.

File: modules/git.py [hunk 17]
```diff
@@ -195,7 +194,6 @@ def do_POST_unsafe(self, data):
         if 'GitHub' in self.headers['User-Agent']:
             event = self.headers['X-Github-Event']
             user = data['sender']['login']
-
```
Comment (sushain97) [inline modules/git.py]: This newline was nice

File: modules/git.py [hunk 18]
```diff
@@ -210,8 +208,11 @@ def do_POST_unsafe(self, data):
                         event_types = value
             else:
                 event_types = None
-
-            if (event_types is not None) and (event not in event_types):
+            event_in_config = False
+            for event_type in event_types:
+                if (event + '_' + data['action'] == event_type) or (event == event_type):
+                     event_in_config = True
```
Comment (sushain97) [inline modules/git.py]: Newline here would be nice.

**Comment by sushain97** [comment]
`['issues_opened', 'issues_closed', 'issues_reopened', 'push', 'issue_comment_created', 'issue']`

Why is `issue` required here? If `closed` is an action for the `issue` webhook, I'd prefer we use `issue_opened`, not `issues_opened`.

**Comment by sushain97** [comment]
Thanks for simplifying the logic though! This is much nicer.

**Comment by avyayv** [comment]
['issues_opened', 'issues_closed', 'issues_reopened', 'push', 'issue_comment_created', issue]

Actually, that was a mistake in my config file that i forgot to fix; the webhook is actually 'issues'; I was testing whether 'issue' would work initially, but when I copy-pasted, I forgot to delete it. This is the updated config.

`
git_events = { '*': ['push'], 'apertium/*': ['issues', 'push'], 'apertium/apertium-all': [], 'apertium/apertium-incubator': [], 'apertium/apertium-nursery': [], 'apertium/apertium-staging': [], 'apertium/apertium-trunk': [], 'apertium/apertium-tools': [], 'avyayv/greatcheaprepowithnocode': ['issues_opened', 'issues_closed', 'issues_reopened', 'issue_comment_created', 'issues_assigned'] }
`

If we want all of the issues actions to appear, we just type 'issues' in the config. The same goes for 'issue_comments'

**Comment by sushain97** [comment]
In that case, why are you using issue_comment_created? Also, seems like the build is failing?

**Comment by avyayv** [comment]
> In that case, why are you using issue_comment_created? Also, seems like the build is failing?

I think 'issue_comment' is a seperate webook from 'issues'

**Comment by sushain97** [comment]
You said it was issue_comments above.

**Comment by avyayv** [comment]
> You said it was issue_comments above.

Whoops, i meant 'issue_comment', ill fix that :\

**Comment by avyayv** [comment]
```
    git_events = { '*': ['push'], 
    'apertium/*': ['issues', 'push'], 
    'apertium/apertium-all': [],
    'apertium/apertium-incubator': [], 
    'apertium/apertium-nursery': [], 
    'apertium/apertium-staging': [], 
    'apertium/apertium-trunk': [],
    'apertium/apertium-tools': [], 
    'avyayv/greatcheaprepowithnocode': ['issues_opened', 'issues_closed', 'issues_reopened', 
    'issue_comment_created'] }
```

**Comment by sushain97** [comment]
git_events = {
        '*': ['push'], 
        'apertium/*': ['push', 'issues_opened', 'issues_closed', 'issues_reopened', 
        'issue_comment_created'], 
        'apertium/apertium-all': [],
        'apertium/apertium-incubator': [], 
        'apertium/apertium-nursery': [], 
        'apertium/apertium-staging': [], 
        'apertium/apertium-trunk': [],
        'apertium/apertium-tools': []
    }
